Add external cloud provider support for OpenStack.

ainmosni opened the pull request:
**What this PR does / why we need it**:
This PR adds external cloud provider support for OpenStack, to accomplish this, the following changes and additions have been made:
    
    * Add a node labeller operator to add distribution labels to nodes.
    * Add Cluster.Spec.Features to the API.
    * Add the FeatureNameExternalCloudProvider feature flag.
    * Create the external cloud provider deployment to the user cluster
      namespace.
    * Add the CSI addon to the addons.
    * Mount the cloud-config into the usercluster controller.
    * Inject the cloud-config into the user cluster.
    * Add a MajorMinor function to semver.

**Does this PR introduce a user-facing change?**:
```release-note
Add support for the external cloud provider for OpenStack.
```

ainmosni committed:
Add external cloud provider support for OpenStack.

This commit adds external cloud provider support for OpenStack, to
accomplish this, the following changes and additions have been made:

* Add a node labeller operator to add distribution labels to nodes.
* Add Cluster.Spec.Features to the API.
* Add the FeatureNameExternalCloudProvider feature flag.
* Create the external cloud provider deployment to the user cluster
  namespace.
* Add the CSI addon to the addons.
* Mount the cloud-config into the usercluster controller.
* Inject the cloud-config into the user cluster.
* Add a MajorMinor function to semver.

Signed-off-by: Daniël Franke <daniel@ams-sec.org>

ainmosni commented:
Note that this is untested as of this writing, but I'm submitting it now so that it can be looked at while I test.

alvaroaleman commented on api/pkg/controller/addon/addon_controller.go:
```diff
@@ -294,6 +294,7 @@ func (r *Reconciler) getAddonManifests(log *zap.SugaredLogger, addon *kubermatic
 		Variables:    make(map[string]interface{}),
 		Cluster:      cluster,
 		Credentials:  credentials,
+		MinorVersion: cluster.Spec.Version.MajorMinor(),
```
I guess this should be the minor version?

ainmosni replied:
No, I should rename that. I need the combo of MajorMinor. :)

alvaroaleman commented on api/pkg/controller/nodelabel/resources/label.go:
```diff
@@ -0,0 +1,23 @@
+	ContainerLinuxLabelValue = "container-linux"
+)
+
+// OSLabelMatchValues is a mapping between OS labels and the strins to match on in OSImage.
```
nit: `s/strins/strings`

alvaroaleman commented on api/pkg/controller/nodelabel/resources/label.go:
```diff
@@ -0,0 +1,23 @@
+package resources
```
Could you move the controller below `pkg/controller/user-cluster-controller-manager`? We want to arrange the controllers by binary they run in, because the current structure of having one folder with all controller confuses ppl

ainmosni replied:
Yeah no problem. :)

alvaroaleman commented on api/pkg/controller/openshift/testdata/Kubermatic_API_image_is_overwritten-usercluster-controller.golden.yaml:
```diff
@@ -41,7 +42,7 @@ spec:
         - "1"
         - -command
-        - '{"command":"/usr/local/bin/user-cluster-controller-manager","args":["-kubeconfig","/etc/kubernetes/kubeconfig/kubeconfig","-metrics-listen-address","0.0.0.0:8085","-health-listen-address","0.0.0.0:8086","-namespace","$(NAMESPACE)","-ca-cert","/etc/kubernetes/pki/ca/ca.crt","-ca-key","/etc/kubernetes/pki/ca/ca.key","-cluster-url","https://test-cluster.alias-europe-west3-c.dev.kubermatic ...
+        - '{"command":"/usr/local/bin/user-cluster-controller-manager"}'
```
That doesn't look right :thinking:

ainmosni replied:
Nice catch, will look into it.

alvaroaleman commented on api/pkg/resources/usercluster/deployment.go:
```diff
@@ -138,7 +139,6 @@ func DeploymentCreator(data userclusterControllerData, openshift bool) reconcili
 					Name:    name,
 					Image:   data.KubermaticAPIImage() + ":" + resources.KUBERMATICCOMMIT,
 					Command: []string{"/usr/local/bin/user-cluster-controller-manager"},
-					Args:    args,
```
This must stay :)

ainmosni replied:
Indeed, I think I made a mistake during rebasing.

alvaroaleman commented on api/pkg/controller/nodelabel/node_label_controller.go:
```diff
@@ -0,0 +1,126 @@
+		return err
+	}
+
+	predicates := predicate.Funcs{
```
You could probably use the same predicate as in the other node-label controller in `controller/user-cluster-controller-manager/node-labeler/node_labeleler.go`. And maybe just combine the two as they are very similiar?

ainmosni replied:
Yeah when I saw that there was another node labeler, I was thinking of combining them but it was a yak I didn't want to shave today.

alvaroaleman commented on api/pkg/controller/user-cluster-controller-manager/resources/reconciler.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 
+	"github.com/kubermatic/kubermatic/api/pkg/controller/usercluster/resources/cloudcontroller"
+
```
Nit: Please put it into the same inport group as the other kubermatic resources

alvaroaleman commented on api/pkg/controller/usercluster/resources/cloudcontroller/clusterrolebinding.go:
```diff
@@ -0,0 +1,32 @@
+				// Can probably be tightened up a bit but for now I'm following the documentation.
+				// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/
+				Name:     "cluster-admin",
+				Kind:     "ClusterROle",
```
`s/ClusterROle/ClusterRole`

alvaroaleman commented on api/pkg/controller/usercluster/resources/cloudcontroller/secret.go:
```diff
@@ -0,0 +1,22 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+// CloudConfig generates the
```
Please finish the sentence or completely remove it

ainmosni replied:
Oops my bad.

alvaroaleman commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -333,6 +333,26 @@ func getApiserverFlags(data *resources.TemplateData, etcdEndpoints []string, ena
 	return flags, nil
 }
 
+func GetKubernetesCloudProviderName(cluster *kubermaticv1.Cluster) string {
```
Please extend `pkg/resources/data.go#GetKubernetesCloudProviderName`  instead

ainmosni replied:
👍

alvaroaleman commented on api/pkg/resources/cloudcontroller/deployment.go:
```diff
@@ -0,0 +1,22 @@
+
+	return func() (name string, create reconciling.DeploymentCreator) {
+		return osName, func(dep *appsv1.Deployment) (*appsv1.Deployment, error) {
+			return &appsv1.Deployment{}, fmt.Errorf("unsupported external cloud controller")
```
Please just return nil for the Deployment and use `errors.New` if there is no format string

ainmosni replied:
👍

alvaroaleman commented on api/pkg/resources/cloudcontroller/openstack.go:
```diff
@@ -0,0 +1,161 @@
+	return flags
+}
+
+func getOSVersion(data *resources.TemplateData) (string, error) {
```
If this is not supported for all versions, can we maybe export a `ExternalCloudControllerSupported(c *kubermaticv1.Cluster) bool` and only if both that func returns true and the feature gate is set enable this?

ainmosni replied:
Good idea.

alvaroaleman commented on api/pkg/resources/controllermanager/deployment.go:
```diff
@@ -224,7 +224,11 @@ func getFlags(cluster *kubermaticv1.Cluster) ([]string, error) {
 	}
 	if cluster.Spec.Cloud.Openstack != nil {
-		flags = append(flags, "--cloud-provider", "openstack")
+		cloudProvider := "openstack"
```
Can we maybe re-use `data.GetKubernetesCloudProviderName` here to not have this in multiple places?

ainmosni replied:
I was thinking that, but that function was still private and in a different package when I worked on this.

alvaroaleman reviewed (commented):
Should we automatically set the feature gate for clusters that support it on create requests?

alvaroaleman commented on api/cmd/user-cluster-controller-manager/main.go:
```diff
@@ -142,6 +145,14 @@ func main() {
+	if runOp.cloudConfig != "" {
+		cloudConfig, err = ioutil.ReadFile(runOp.cloudConfig)
+		if err != nil {
+			glog.Fatalf("Failed to read cloud-config: %v", err)
```
This will fail, we don't use `glog` anymore but `zap`, there should be a `log` variable you can use

ainmosni replied:
I do use the `log` facility, wonder how you still saw that. :)

alvaroaleman commented on api/pkg/addon/template.go:
```diff
@@ -35,6 +35,7 @@ func txtFuncMap(overwriteRegistry string) template.FuncMap {
 type TemplateData struct {
 	Addon        *kubermaticv1.Addon
 	Kubeconfig   string
+	MinorVersion string
```
Can we rename this here as well?

ainmosni replied:
Already renamed, looks like you were reviewing an older version.

alvaroaleman commented on api/pkg/controller/cluster/resources.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/kubermatic/kubermatic/api/pkg/resources/cloudcontroller"
```
Nit: Please group this with the other kuberamatic imports

ainmosni replied:
Serves me right for letting VSCode import for me and forgetting to organise it. Fixed locally, will push it in a bit.

alvaroaleman commented on api/pkg/controller/user-cluster-controller-manager/nodelabel/node_label_controller.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func (r *Reconciler) applyLabel(ctx context.Context, node *corev1.Node, label string) error {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Coukd you use Patch instead?  That imposes less load on the APIserver and is pretty much what we want here. We couldn't use it before because our controller-runtime version didn't support it but thats now the case, check e.G. pkg/controller/user-cluster-controller-manager/node-labeler/node_labeleler.go:115

ainmosni replied:
Sure, would it replace the whole `RetryOnConflict` or just the `r.Client.Update` ?

alvaroaleman replied:
The whole `RetryOnConflict`, just one attempt on patching.

`RetryOnConflict` with the modifiers is essentially a poor persons patching :grimacing:

ainmosni replied:
Thanks for the tip, will keep that in mind. :)

alvaroaleman commented on api/pkg/resources/cloudcontroller/openstack.go:
```diff
@@ -0,0 +1,161 @@
+	}
+}
+
+func getOSPodAnnotations(data *resources.TemplateData) map[string]string {
```
I think we can just remove this

alvaroaleman commented on api/pkg/resources/controllermanager/deployment.go:
```diff
@@ -211,59 +211,41 @@ func getFlags(cluster *kubermaticv1.Cluster) ([]string, error) {
-		if cluster.Spec.Version.Semver().Minor() >= 15 {
-			// Required so multiple clusters using the same resource group can
-			// allocate public IPs. Ref: https://github.com/kubernetes/kubernetes/pull/77630
-			flags = append(flags, "--cluster-name", cluster.Name)
```
This got lost or am I missing it?

ainmosni replied:
I shouldn't do big deletions when super tired. :)

alvaroaleman reviewed (commented):
Some nits, but looks mostly good. Do you want to add anything to enable this by default on new clusters that support it?

alvaroaleman commented:
/lgtm
/approve

Since this won't be used by default I am fine merging this as it is

alvaroaleman commented:
The unittests fail with a:
```
--- FAIL: TestRetagImageForAllVersions (0.60s)
    main_test.go:31: Error calling getImagesForVersion: failed to get images from addons: failed to get images for addon cinder-csi-plugin: failed to parse addon templates in ../../../addons/cinder-csi-plugin: failed to parse file ../../../addons/cinder-csi-plugin/nodeplugin.yaml: template: nodeplugin.yaml:61: undefined variable "$version"
    main_test.go:31: Error calling getImagesForVersion: failed to get images from addons: failed to get images for addon cinder-csi-plugin: failed to parse addon templates in ../../../addons/cinder-csi-plugin: failed to parse file ../../../addons/cinder-csi-plugin/nodeplugin.yaml: template: nodeplugin.yaml:61: undefined variable "$version"
    main_test.go:31: Error calling getImagesForVersion: failed to get images from addons: failed to get images for addon cinder-csi-plugin: failed to parse addon templates in ../../../addons/cinder-csi-plugin: failed to parse file ../../../addons/cinder-csi-plugin/nodeplugin.yaml: template: nodeplugin.yaml:61: undefined variable "$version"
    main_test.go:31: Error calling getImagesForVersion: failed to get images from addons: failed to get images for addon cinder-csi-plugin: failed to parse addon templates in ../../../addons/cinder-csi-plugin: failed to parse file ../../../addons/cinder-csi-plugin/nodeplugin.yaml: template: nodeplugin.yaml:61: undefined variable "$version"
    main_test.go:31: Error calling getImagesForVersion: failed to get images from addons: failed to get images for addon cinder-csi-plugin: failed to parse addon templates in ../../../addons/cinder-csi-plugin: failed to parse file ../../../addons/cinder-csi-plugin/nodeplugin.yaml: template: nodeplugin.yaml:61: undefined variable "$version"
```

alvaroaleman commented:
/test all

alvaroaleman commented on addons/default-storage-class/storage-class.yaml:
```diff
@@ -52,8 +52,13 @@ metadata:
   labels:
     kubernetes.io/cluster-service: "true"
   name: standard
+{{ if (.Cluster.Spec.Features.externalCloudProvider) }}
```
Storageclasses are immutable, so can we use complete different storage classes?

ainmosni replied:
We could, but as new clusters with the external CCM enabled should only use the CSI storage, mutating the standard class seemed more logical.

alvaroaleman replied:
True, but we eventually want to have a migration and at that point in time we will have to rename it. And if we keep the name now that would mean that even clusters that never used the in-tree provider will end up with two identical but differently named storageclasses, which is IMHO confusing

ainmosni replied:
Fair enough, I'll change this.

ainmosni replied:
Changed to create a storage class named `cinder-csi` when the external ccm is enabled.

alvaroaleman commented on api/pkg/resources/cloudcontroller/openstack.go:
```diff
@@ -0,0 +1,167 @@
+
+func getOSVersion(data *resources.TemplateData) (string, error) {
+	x := data.Cluster().Spec.Version.String()
+	switch {
```
I think we can just use `data.Cluster().Spec.Version.Minor()`, no reason to do string prefix comparisons

ainmosni replied:
We could, until K8s 2 comes around. :)

alvaroaleman replied:
Right, but that is not going to happen anytime soon, so currently this is solving a problem we don't have. Once a 2.X exists (for which upstream doesn't even have plans currently) we will have to update a variety of places anyways

ainmosni replied:
True, at the very least I could use the `MajorMinor` function that I created after I wrote this code so I can do plain equality instead of prefix comparison.

ainmosni replied:
Adapted it to use the `MajorMinor` function, which cleans it up a bit.

alvaroaleman commented on addons/cinder-csi-plugin/controllerplugin-rbac.yaml:
```diff
@@ -0,0 +1,212 @@
+{{ if (.Cluster.Spec.Features.externalCloudProvider) }}
```
Can we move this into the `default-storage-class` addon? With the current approach we'd have to activate this addon by default which will be confusing for setups where Openstack is unsed

ainmosni replied:
Yeah I thought about that, I thought that adding it to another addon would make it less clear what the addon did. Of course, we could rename `default-storage-class` to just `storage` and it would be more clear.

alvaroaleman replied:
I do agree that the name would be clearer, but if we did that, we would re-create all resources in it for existing clusters so that is not an option I am afraid :grimacing: Compatibility can be annoying sometimes

ainmosni replied:
I could also just rename `cinder-csi-plugin` to just `csi` or just `storage`, would look a lot less jarring, even if it is a noop on non-OpenStack clusters for now.

alvaroaleman replied:
Sgtm

alvaroaleman commented on api/pkg/resources/machinecontroller/deployment.go:
```diff
@@ -114,12 +114,17 @@ func DeploymentCreatorWithoutInitWrapper(data machinecontrollerData) reconciling
 			}
 
+			externalCloudProvider := false
+			if x := data.Cluster().Spec.Features[resources.FeatureNameExternalCloudProvider]; x {
```
Can be simplified as `externalCloudProvider := data.Cluster().Spec.Features[resources.FeatureNameExternalCloudProvider]`

ainmosni replied:
Done, confused Go's behaviour with other languages where you'd get a "unknown key" error. :)

alvaroaleman commented:
/test all

alvaroaleman commented:
/test all

alvaroaleman commented:
/test pull-kubermatic-e2e-openstack-coreos-1.15

ainmosni commented:
@alvaroaleman It's kinda a shame that I can't see the test results myself, then I could see what's breaking that one test.

ainmosni committed:
Upgrade CSI sidecards.

For details see: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-dev/8QqL4YcI1cM/tRe1D2U9AgAJ

Signed-off-by: Daniël Franke <daniel@ams-sec.org>

alvaroaleman commented:
> Chart kubermatic was modified but its version (1.0.18) was not changed. Please adjust config/kubermatic/Chart.yaml.

/test all

ainmosni committed:
Bump chart.

Signed-off-by: Daniël Franke <daniel@ams-sec.org>

ainmosni commented:
Ok, bumped the chart. :)

alvaroaleman commented:
/test all

alvaroaleman commented:
/lgtm
/approve

alvaroaleman reviewed (approved):
/lgtm

alvaroaleman commented:
/retest

alvaroaleman commented:
/lgtm
/test all

ainmosni committed:
Update test fixtures.

Signed-off-by: Daniël Franke <daniel@ams-sec.org>

alvaroaleman commented:
/lgtm

alvaroaleman commented:
/retest

alvaroaleman commented:
/retest

alvaroaleman commented:
/retest
